apollo-router/src/plugins/cache/invalidation.rs

```diff
@@ -128,117 +88,130 @@ async fn start(
-    let key_prefix = request.key_prefix();
-    let subgraph = request.subgraph_name();
-    tracing::debug!(
-        "got invalidation request: {request:?}, will scan for: {}",
-        key_prefix
-    );
-
-    let mut stream = storage.scan(key_prefix.clone(), Some(scan_count));
-    let mut count = 0u64;
-    let mut error = None;
+    async fn handle_request(
+        &self,
+        redis_storage: &RedisCacheStorage,
+        origin: &'static str,
+        request: &InvalidationRequest,
+    ) -> Result<u64, InvalidationError> {
+        let key_prefix = request.key_prefix();
+        let subgraph = request.subgraph_name();
+        tracing::debug!(
+            "got invalidation request: {request:?}, will scan for: {}",
+            key_prefix
+        );
 
-    while let Some(res) = stream.next().await {
-        match res {
-            Err(e) => {
-                tracing::error!(
-                    pattern = key_prefix,
-                    error = %e,
-                    message = "error scanning for key",
-                );
-                error = Some(e);
-                break;
-            }
-            Ok(scan_res) => {
-                if let Some(keys) = scan_res.results() {
-                    let keys = keys
-                        .iter()
-                        .filter_map(|k| k.as_str())
-                        .map(|k| RedisKey(k.to_string()))
-                        .collect::<Vec<_>>();
-                    if !keys.is_empty() {
-                        count += storage.delete(keys).await.unwrap_or(0) as u64;
+        let mut stream = redis_storage.scan(key_prefix.clone(), Some(self.scan_count));
+        let mut count = 0u64;
+        let mut error = None;
 
-                        u64_counter!(
-                            "apollo.router.operations.entity.invalidation.entry",
-                            "Entity cache counter for invalidated entries",
-                            count,
-                            "origin" = origin,
-                            "subgraph.name" = subgraph.clone()
-                        );
+        while let Some(res) = stream.next().await {
+            match res {
+                Err(e) => {
+                    tracing::error!(
+                        pattern = key_prefix,
+                        error = %e,
+                        message = "error scanning for key",
+                    );
+                    error = Some(e);
+                    break;
+                }
+                Ok(scan_res) => {
+                    if let Some(keys) = scan_res.results() {
+                        let keys = keys
+                            .iter()
+                            .filter_map(|k| k.as_str())
+                            .map(|k| RedisKey(k.to_string()))
+                            .collect::<Vec<_>>();
+                        if !keys.is_empty() {
+                            let deleted = redis_storage.delete(keys).await.unwrap_or(0) as u64;
```

Comment:
In my branch when I tested to improve perfs I put all the keys to delete in an array to not block the scan and then delete everything once. What do you think ?

Reply:
In what way would this block the scan?